File: 002_Variable_Data_Types_Operators/1_sizeof.cpp
#include <iostream>  // Header for input-output operations
using namespace std;

int main() {
    // ----------------------------------------------
    // 🔹 Using sizeof with data types directly
    // ----------------------------------------------
    cout << "Size of int: " << sizeof(int) << " bytes" << endl;      // Typically 4 bytes
    cout << "Size of float: " << sizeof(float) << " bytes" << endl;  // Typically 4 bytes
    cout << "Size of double: " << sizeof(double) << " bytes" << endl;// Typically 8 bytes
    cout << "Size of char: " << sizeof(char) << " byte" << endl;     // Always 1 byte
    cout << "Size of bool: " << sizeof(bool) << " byte" << endl;     // Usually 1 byte

    // ----------------------------------------------
    // 🔹 Declaring constants using const keyword
    // ----------------------------------------------
    // const means the variable's value cannot be changed after assignment
    const int a = 0;
    const float b = 0.0f;
    const double c = 0.0;
    const char d = 'A';
    const bool e = true;

    // ----------------------------------------------
    // 🔹 Using sizeof with constant variables
    // ----------------------------------------------
    cout << "Size of int (const a): " << sizeof(a) << " bytes" << endl;
    cout << "Size of float (const b): " << sizeof(b) << " bytes" << endl;
    cout << "Size of double (const c): " << sizeof(c) << " bytes" << endl;
    cout << "Size of char (const d): " << sizeof(d) << " byte" << endl;
    cout << "Size of bool (const e): " << sizeof(e) << " byte" << endl;

    return 0;
}

/*
-----------------------------------------------
🔍 Notes:

1. sizeof(type) ➝ Tells the memory (RAM) size occupied by a data type.
   Example: sizeof(int) usually returns 4.

2. sizeof(variable) ➝ Also works the same way.
   Example: sizeof(a) is same as sizeof(int) if a is of type int.

3. const keyword ➝ Used to make variables constant (read-only).
   You can still use sizeof on const variables.

4. Sizes of data types may vary depending on:
   - System architecture (32-bit vs 64-bit)
   - Compiler being used

5. char and bool are usually 1 byte.
   double is often 8 bytes, float is 4 bytes.

6. In Python, we use `sys.getsizeof()` which includes extra overhead (not just data).
-----------------------------------------------
*/

File: 002_Variable_Data_Types_Operators/1.sizeof.py

import sys  # Required to use sys.getsizeof()

# -----------------------------------------------
# 🔹 Declaring variables of different data types
# -----------------------------------------------
a = 0           # Integer
b = 3.14        # Float (decimal number)
c = 'A'         # Character (actually a string of length 1 in Python)
d = True        # Boolean (True/False)
e = "Shreya"    # String (sequence of characters)

# -----------------------------------------------
# 🔹 Printing memory size using sys.getsizeof()
# -----------------------------------------------
print("Size of int:", sys.getsizeof(a), "bytes")        # Includes internal overhead
print("Size of float:", sys.getsizeof(b), "bytes")
print("Size of char:", sys.getsizeof(c), "bytes")       # 1-char string
print("Size of bool:", sys.getsizeof(d), "bytes")
print("Size of string:", sys.getsizeof(e), "bytes")     # Size increases with length

"""
-----------------------------------------------
📝 Notes:

1. In C++:
   - sizeof(type) gives exact memory size of data type (e.g., int = 4 bytes).
   - It's fixed for a given system/architecture.

2. In Python:
   - sys.getsizeof() returns the size of an **object in memory**.
   - It includes both:
     ✔ actual data size
     ✔ internal interpreter overhead

3. For example:
   - An empty string uses memory too (due to internal object structure).
   - A string’s size grows as the number of characters increases.

4. Data Types Mapped:
   - `int` ➝ Integer (like C++)
   - `float` ➝ Floating-point number
   - `str` ➝ String (char is just a 1-length string in Python)
   - `bool` ➝ Boolean (True/False)

5. Output may vary slightly based on:
   - Python version
   - Operating System
   - System architecture (32-bit/64-bit)

-----------------------------------------------
"""